### DISCUSSION TITLE
Run until complete in thread / nested event loops

### DISCUSSION THREAD

**Discussion by davidbrochart** [agronholm/anyio#341] [Q&A] [answered]
It would be nice to have the "opposite" of `anyio.to_thread.run_sync()`, which would run an async function until complete in a thread with its own event loop. Something like `anyio.to_thread.run_async()`:
```python
from anyio import to_thread

async def foo():
    print('Hello, world!')

def main():
    to_thread.run_async(foo)

main()
```
It would help to solve the problem where you want to run an async function in a blocking way but you don't know if you are being called from an already running event loop, so you cannot run until complete. An alternative is to use [nest-asyncio](https://github.com/erdewit/nest_asyncio), but it has issues too.

Maybe it could also be made transparent, such as the following code works:
```python
from anyio import run

async def foo():
    print('Hello, world!')

async def main():
    run(foo)

run(main)
```
This currently gives a `RuntimeError: Already running asyncio in this thread`.

**Comment by graingert** [comment]
there's https://anyio.readthedocs.io/en/stable/api.html#anyio.from_thread.start_blocking_portal

**Comment by davidbrochart** [comment, marked as answer]
Just in case someone reads this issue, this is the way. Thanks!
```python
from anyio import run, from_thread

async def foo():
    print('Hello, world!')

async def main():
    with from_thread.start_blocking_portal() as portal:
        portal.call(foo)

run(main)
```

**Reply by davidbrochart** [reply]
As its name implies, `start_blocking_portal` will block. What if I want to run a new event loop in a separate thread, as a task in the event loop thread? This seems to work but I'm not sure it's the correct way:
```py
from functools import partial
from anyio import create_task_group, run, sleep, to_thread

async def thread():
    while True:
        await sleep(1)
        print("From thread")

async def main():
    async with create_task_group() as tg:
        tg.start_soon(to_thread.run_sync, partial(run, thread))
        while True:
            await sleep(1)
            print("From main")

run(main)
```

**Reply by gschaffner** [reply]
> Just in case someone reads this issue, this is the way. Thanks!
> 
> ...

fyi, this currently only works if using `anyio.run(main, backend="asyncio")`. it's broken on trio; see #523 and #525.
